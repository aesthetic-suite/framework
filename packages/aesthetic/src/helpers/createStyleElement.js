/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

export default function createStyleElement(id: string): HTMLElement {
  const style = document.createElement('style');
  style.type = 'text/css';
  style.id = `aesthetic-${id}`;

  // $FlowIgnore We know the document exists
  document.head.appendChild(style);

  return style;
}
