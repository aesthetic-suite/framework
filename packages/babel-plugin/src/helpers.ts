import { types as t } from '@babel/core';
import { createDebugger } from '@boost/debug';

export const debug = createDebugger('aesthetic:babel');

export function isFunction(node: t.Node) {
  return t.isFunctionExpression(node) || t.isArrowFunctionExpression(node);
}
