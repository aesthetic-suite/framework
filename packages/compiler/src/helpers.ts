import { ColorShade } from '@aesthetic/system';

export function formatShade(value: number | string): ColorShade {
	return String(value).padStart(2, '0').slice(0, 2) as ColorShade;
}

export function formatUnit(value: number): string {
	return value.toFixed(2).replace('.00', '');
}
