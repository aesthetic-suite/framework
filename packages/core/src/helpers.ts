import { Theme } from '@aesthetic/system';
import { objectLoop } from '@aesthetic/utils';
import { SheetParams } from './types';

export function createCacheKey(params: SheetParams, type: string): string {
  let key = type;

  // Since all other values are scalars, we can just join the values.
  // This is 3x faster than JSON.stringify(), and 1.5x faster than Object.values()!
  objectLoop(params, (value) => {
    key += value;
  });

  return key;
}

export function createDefaultParams(theme: Theme<{}>, params: SheetParams): Required<SheetParams> {
  return {
    contrast: theme.contrast,
    deterministic: false,
    direction: 'ltr',
    scheme: theme.scheme,
    theme: theme.name,
    unit: 'px',
    vendor: false,
    ...params,
  };
}
