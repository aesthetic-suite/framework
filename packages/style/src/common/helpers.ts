/* eslint-disable no-magic-numbers */

import { Value, MaybeValue, CSS, Sheet } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';
import { IMPORT_RULE, STYLE_RULE } from './constants';

export function insertRule(sheet: Sheet, rule: CSS, index?: number): number {
  try {
    return sheet.insertRule(rule, index ?? sheet.cssRules.length);
  } catch {
    // Vendor prefixed properties, pseudos, etc, that are inserted
    // into different vendors will trigger a failure. For example,
    // `-moz` or `-ms` being inserted into WebKit.
    // There's no easy way around this, so let's just ignore the
    // error so that subsequent styles are inserted.
    // istanbul ignore next
    return -1;
  }
}

export function insertAtRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // At-rules must be inserted before normal style rules.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type === STYLE_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function insertImportRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // Import rules must be inserted at the top of the style sheet,
  // but we also want to persist the existing order.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type !== IMPORT_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import';
}

export function isMediaRule(value: string): boolean {
  return value.slice(0, 6) === '@media';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

export function isSupportsRule(value: string): boolean {
  return value.slice(0, 9) === '@supports';
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      // eslint-disable-next-line no-console
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}

export function joinQueries(prev: string | undefined, next: string): string {
  if (prev) {
    return `${prev} and ${next}`;
  }

  return next;
}
