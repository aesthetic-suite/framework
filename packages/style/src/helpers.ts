/* eslint-disable no-magic-numbers */

import { Value, MaybeValue } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import' && value[7] === ' ';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}
