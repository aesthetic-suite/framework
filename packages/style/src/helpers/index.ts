import createAtomicCacheKey from './createAtomicCacheKey';
import formatConditions from './formatConditions';
import formatDeclaration from './formatDeclaration';
import formatDeclarationBlock from './formatDeclarationBlock';
import formatRule from './formatRule';
import formatVariable from './formatVariable';
import formatVariableName from './formatVariableName';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isAtRule from './isAtRule';
import isFontFaceFule from './isFontFaceFule';
import isImportRule from './isImportRule';
import isInvalidValue from './isInvalidValue';
import isKeyframesRule from './isKeyframesRule';
import isMediaRule from './isMediaRule';
import isNestedSelector from './isNestedSelector';
import isSupportsRule from './isSupportsRule';
import isUnitlessProperty from './isUnitlessProperty';
import isVariable from './isVariable';

export {
  createAtomicCacheKey,
  formatConditions,
  formatDeclaration,
  formatDeclarationBlock,
  formatRule,
  formatVariable,
  formatVariableName,
  getDocumentStyleSheet,
  isAtRule,
  isFontFaceFule,
  isImportRule,
  isInvalidValue,
  isKeyframesRule,
  isMediaRule,
  isNestedSelector,
  isSupportsRule,
  isUnitlessProperty,
  isVariable,
};
