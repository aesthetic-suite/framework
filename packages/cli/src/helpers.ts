import { CONFIG_FOLDER } from '@aesthetic/compiler';
import { Path } from '@boost/common';
import { CWD } from './constants';

export function getConfigFolderDir(name: string): Path {
  return new Path(CWD, CONFIG_FOLDER, name);
}

export function validateSystemName(value: string) {
  if (value && !value.match(/^[-a-z0-9]+$/giu)) {
    throw new Error('Name may only contain numbers, letters, and dashes.');
  }
}
