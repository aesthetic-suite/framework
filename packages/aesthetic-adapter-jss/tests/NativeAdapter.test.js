/* eslint-disable sort-keys */

import { create } from 'jss';
import preset from 'jss-preset-default';
import JSSAdapter from '../src/NativeAdapter';
import {
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_NATIVE_PARTIAL,
} from '../../../tests/mocks';
import { renderJSSStyles } from '../../../tests/helpers';

describe('aesthetic-adapter-jss/NativeAdapter', () => {
  let instance;

  beforeEach(() => {
    const jss = create();
    jss.setup(preset());

    instance = new JSSAdapter(jss);
  });

  it('can customize the JSS instance through the constructor', () => {
    const jss = create();
    instance = new JSSAdapter(jss);

    expect(instance.jss).toBe(jss);
  });

  it('transforms style declarations into class names', () => {
    expect(instance.transform(instance.create(SYNTAX_NATIVE_PARTIAL).button)).toBe('button-0-3-1');
  });

  it('combines different style declarations into unique class names', () => {
    const sheet = instance.create({
      foo: {
        color: 'red',
        display: 'block',
      },
      bar: {
        color: 'green',
        margin: 5,
      },
      baz: {
        color: 'blue',
        padding: 5,
      },
    });

    expect(instance.transform(sheet.foo)).toBe('foo-0-4-1');
    expect(instance.transform(sheet.bar)).toBe('bar-0-4-2');
    expect(instance.transform(sheet.baz)).toBe('baz-0-4-3');
    expect(instance.transform(sheet.foo, sheet.baz)).toBe('foo-0-4-1 baz-0-4-3');
    expect(instance.transform(sheet.bar, sheet.foo)).toBe('bar-0-4-2 foo-0-4-1');
  });

  it('handles pseudo selectors', () => {
    expect(
      instance.transform(
        instance.create({
          pseudo: {
            position: 'fixed',
            '&:hover': {
              position: 'static',
            },
            '&::before': {
              position: 'absolute',
            },
          },
        }).pseudo,
      ),
    ).toBe('pseudo-0-5-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles fallbacks', () => {
    expect(
      instance.transform(
        instance.create({
          fallback: {
            background: 'red',
            display: 'flex',
            fallbacks: [
              { background: 'linear-gradient(...)' },
              { display: 'flex-box' },
              { display: 'box' },
            ],
          },
        }).fallback,
      ),
    ).toBe('fallback-0-6-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles font faces', () => {
    expect(
      instance.transform(
        instance.create({
          '@font-face': FONT_ROBOTO_FLAT_SRC,
          font: {
            fontFamily: 'Roboto',
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('font-0-7-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles animations', () => {
    expect(
      instance.transform(
        instance.create({
          '@keyframes fade': KEYFRAME_FADE,
          animation: {
            animationName: 'fade',
            animationDuration: '3s, 1200ms',
            animationIterationCount: 'infinite',
          },
        }).animation,
      ),
    ).toBe('animation-0-8-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles media queries', () => {
    expect(
      instance.transform(
        instance.create({
          media: {
            color: 'red',
            '@media (min-width: 300px)': {
              color: 'blue',
            },
            '@media (max-width: 1000px)': {
              color: 'green',
            },
          },
        }).media,
      ),
    ).toBe('media-0-9-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles supports', () => {
    expect(
      instance.transform(
        instance.create({
          sup: {
            display: 'block',
            '@supports (display: flex)': {
              display: 'flex',
            },
            '@supports not (display: flex)': {
              float: 'left',
            },
          },
        }).sup,
      ),
    ).toBe('sup-0-10-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('prefixes class names with style name', () => {
    expect(instance.transform(instance.create(SYNTAX_NATIVE_PARTIAL, 'prefix').button)).toBe(
      'prefix-button-0-11-1',
    );
  });

  it('can transform dynamic styles', () => {
    expect(
      instance.transform({
        width: 10,
        height: 10,
      }),
    ).toBe('dynamic-inline0-0-12-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles descendant selectors', () => {
    expect(
      instance.transform(
        instance.create({
          list: {
            margin: 0,
            padding: 0,
            '&> li': {
              listStyle: 'bullet',
            },
          },
        }).list,
      ),
    ).toBe('list-0-13-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles attribute selectors', () => {
    expect(
      instance.transform(
        instance.create({
          attr: {
            display: 'block',
            '&[disabled]': {
              opacity: 0.5,
            },
          },
        }).attr,
      ),
    ).toBe('attr-0-14-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });

  it('handles multiple font faces', () => {
    expect(
      instance.transform(
        instance.create({
          '@font-face': FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
          font: {
            fontFamily: 'Circular',
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('font-0-15-1');

    expect(renderJSSStyles(instance)).toMatchSnapshot();
  });
});
