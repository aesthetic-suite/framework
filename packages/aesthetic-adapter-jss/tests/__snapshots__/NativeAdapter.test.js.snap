// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aesthetic-adapter-jss/NativeAdapter handles animations 1`] = `"@keyframes fade {from {opacity: 0;}to {opacity: 1;}}.animation-0-8-1 {animation-name: fade;animation-duration: 3s, 1200ms;animation-iteration-count: infinite;}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles descendant selectors 1`] = `".list-0-13-1 {margin: 0;padding: 0;}.list-0-13-1 > li {list-style: bullet;}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles fallbacks 1`] = `".fallback-0-6-1 {background: linear-gradient(...);display: flex-box;display: box;display: flex;background: red;}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles font faces 1`] = `"@font-face {font-family: Roboto;font-style: normal;font-weight: normal;src: local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}.font-0-7-1 {font-size: 20px;font-family: Roboto;}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles media queries 1`] = `".media-0-9-1 {color: red;}@media (min-width: 300px) {.media-0-9-1 {color: blue;}}@media (max-width: 1000px) {.media-0-9-1 {color: green;}}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles pseudo selectors 1`] = `".pseudo-0-5-1 {position: fixed;}.pseudo-0-5-1:hover {position: static;}.pseudo-0-5-1::before {position: absolute;}"`;

exports[`aesthetic-adapter-jss/NativeAdapter handles supports 1`] = `".sup-0-10-1 {display: block;}@supports (display: flex) {.sup-0-10-1 {display: flex;}}@supports not (display: flex) {.sup-0-10-1 {float: left;}}"`;
